### PR TITLE
Support for htmlSafe/SafeString

### LINKED ISSUE #101: Fix {{{foo}}} where foo is an empty string
https://github.com/tildeio/glimmer/blob/master/packages/glimmer-runtime/lib/dom.ts#L89

If bar === “” on initial render (or later become empty), it seems like we won't be able to update it after. Should probably put an empty comment in there.

### PR DESCRIPTION
**Pull request by zackthehuman**
This creates a new type of Append opcode that handles the value when it is a SafeString and also when it is a normal string. For SafeString it enters into "html" mode, and if the input changes into a normal string it switches into "text" mode. It can switch freely between the two modes and handles updates accordingly.

- The opcode also supports const expressions.
- New tests have been added which test the switching behavior of the opcode.
- Checked the code TypeScript errors (via `tslint` and `vscode-build`) and corrected them.

This change appears to fix #101 so I added a test that checks that case specifically.

### CODE REVIEW

**Comment by krisselden** [comment]
rebase and squash

File: packages/glimmer-runtime/lib/syntax/core.ts [hunk 1]
```diff
@@ -233,27 +233,6 @@ export class Append extends StatementSyntax {
   }
 }
 
-class HelperInvocationReference implements Reference<Insertion> {
```
Comment (chancancode) [inline packages/glimmer-runtime/lib/syntax/core.ts]: :+1:

File: packages/glimmer-runtime/lib/environment.ts [hunk 2]
```diff
@@ -200,21 +200,30 @@ export abstract class Environment {
...
 }
 
 export type Insertion = string | SafeString | Node;
+export type TrustedInsertion = string | Node;
 
 type PositionalArguments = Opaque[];
 type KeywordArguments = Dict<Opaque>;
 
 export interface Helper {
-  (args: EvaluatedArgs): PathReference<Opaque>;
```
Comment (chancancode) [inline packages/glimmer-runtime/lib/environment.ts]: is this change intentional?

Comment (zackthehuman) [inline packages/glimmer-runtime/lib/environment.ts]: Yes, otherwise there doesn't seem to be a way to return a `ConstReference` from a `Helper` since they aren't `PathReference`s. Is there a better way to resolve?

Comment (chancancode) [inline packages/glimmer-runtime/lib/environment.ts]: @zackthehuman @krisselden has filled me in on the details 👍 looking into it!

Is 4148235 the latest commit you have? I already rebased on top of it, so I think we are good to merge after resolving this point!

Comment (zackthehuman) [inline packages/glimmer-runtime/lib/environment.ts]: Yes that is the most up-to-date commit.
